sktime/transformations/series/subsequence_extraction.py

```diff
@@ -0,0 +1,194 @@
+
+from sktime.transformations.base import BaseTransformer
+
+__all__ = ["SubsequenceExtractionTransformer"]
+__author__ = ["wirrywoo"]
+
+
+class SubsequenceExtractionTransformer(BaseTransformer):
+    r"""
+    Extract contiguous subsequences of specified length based on rolling aggregates.
+
+    A transformer for the extraction of contiguous subsequences of specified
+    length based on maximal/minimal rolling window aggregates.
+
+    Given a sequence :math:`\\{x_1, x_2, \cdots, x_n \\}` and `subseq_len` integer
+    :math:`k` such that :math:`0 < k \leq n`, the transformer's task is to find index
+    :math:`i` satisfying :math:`1 \leq i \leq i + k - 1 \leq n` such that for given
+    `aggregate_fn` :math:`A: \mathbb{R}^k \longrightarrow \mathbb{R}`:
+
+    1. :math:`A(x_{i}, \cdots, x_{i+k-1})` is maximal when `selector = 'max'`, and
+    2. :math:`A(x_{i}, \cdots, x_{i+k-1})` is minimal when `selector = 'min'`.
+
+    When `aggregate_fn = np.sum` and `selector = 'max'`, the problem degenerates to the
+    `maximum sum subarray problem <https://en.wikipedia.org/wiki/Maximum_subarray_problem>`_.
+
+    Parameters
+    ----------
+    subseq_len : int
+        Length of the subsequence in .iloc units. Must be less than the lengths of all
+        input series.
+    aggregate_fn : np._core._multiarray_umath._ArrayFunctionDispatcher, default: np.sum
+        NumPy callable used to aggregate values in contiguous subsequence to a scalar.
+    kwargs : dict, default: {}
+        Dictionary of additional keyword arguments to pass to aggregate_fn.
+    selector : {'max', 'min'}, default: 'max'
+        Function used to decide which subsequence to return from the set of scalars or
+        primitives.
+
+    Examples
+    --------
+    >>> from sktime.transformations.series.subsequence_extraction import (
+    >>>     SubsequenceExtractionTransformer
+    >>> )
+    >>> from sktime.utils._testing.hierarchical import _make_hierarchical
+    >>> X = _make_hierarchical(same_cutoff=False)
+    >>> subseq_extract = SubsequenceExtractionTransformer(subseq_len = 3)
+    >>> subseq_extract.fit(X)
+    >>> X_transformed = subseq_extract.transform(X)
+
+    References
+    ----------
+    Jon Bentley. 1984. Programming pearls: algorithm design techniques.
+    Commun. ACM 27, 9 (Sept. 1984), 865-873. https://doi.org/10.1145/358234.381162
+    """
+
+    _tags = {
+        "univariate-only": False,
+        "authors": ["wirrywoo"],
+        "maintainers": ["wirrywoo"],
+        "scitype:transform-input": "Series",
+        "scitype:transform-output": "Series",
+        "scitype:instancewise": False,
+        "scitype:transform-labels": "None",
+        "X_inner_mtype": "pd.DataFrame",
+        "fit_is_empty": False,
+        "capability:inverse_transform": False,
+        "capability:unequal_length:removes": True,
+        "handles-missing-data": False,
+    }
+
+    def __init__(self, subseq_len, aggregate_fn=np.sum, kwargs={}, selector="max"):
```

Comment:
we must avoid functions and mutable defaults like the dict.

The pattern to deal with this is using `None`, and internally a private attribute `self._aggregate_fn` (since the attributes of same name should always be identical to init args).